Delete line if suggestion would replace it with an empty line

clubby789 opened the pull request:
Fixes #120296

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7640100371/job/20814509611) [(plain)](https://github.com/rust-lang/rust/commit/020c6ca398d4d5d28d1b3118ba887b83f5894d29/checks/20814509611/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=clubby789
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_fb8c4eaf-669d-4dfe-9e34-7d2f4179b2b9
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=3f86ee4795cd5c1c94f3699c94f2c4c251820627
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_fb8c4eaf-669d-4dfe-9e34-7d2f4179b2b9
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_fb8c4eaf-669d-4dfe-9e34-7d2f4179b2b9
GITHUB_TRIGGERING_ACTOR=clubby789
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/120305/merge
GITHUB_WORKFLOW_SHA=3f86ee4795cd5c1c94f3699c94f2c4c251820627
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
Built container sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Looks like docker image is the same as before, not uploading
https://ci-caches.rust-lang.org/docker/7ebc15c01a233894034d277c8cce4e949f4e7791f66b4727c8fb6e058a0b8171d6152e1441d677cef0653843ceeee469c097b8699b2bb74249e674f6aa1a8813
sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Wed Jan 24 12:24:30 UTC 2024
  network time: Wed, 24 Jan 2024 12:24:30 GMT
  network time: Wed, 24 Jan 2024 12:24:30 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'build.optimized-compiler-builtins', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
diff of fixed:

2 #![deny(unused_imports)]
3 
4 // Check that attributes get removed too. See #87973.
- 
6 //~^ ERROR unused import
8 fn main() {}


The actual fixed differed from the expected fixed.
The actual fixed differed from the expected fixed.
Actual fixed saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/imports/unused-import-issue-87973/unused-import-issue-87973.fixed
To only update this specific test, also pass `--test-args imports/unused-import-issue-87973.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/imports/unused-import-issue-87973.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/imports/unused-import-issue-87973" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/imports/unused-import-issue-87973/auxiliary"
--- stderr -------------------------------
error: unused import: `std::fs`
##[error]  --> /checkout/tests/ui/imports/unused-import-issue-87973.rs:8:5
Build completed unsuccessfully in 0:13:02
```

</details>

petrochenkov commented on compiler/rustc_resolve/src/check_unused.rs:
```diff
@@ -373,7 +373,14 @@ impl Resolver<'_, '_> {
+                // If this is the only import on the line, remove the whole line
+                if let Ok(next) = tcx.sess.source_map().span_to_next_source(*sp) {
+                    let whitespace = next.chars().take_while(char::is_ascii_whitespace).count();
+                    *sp = sp.with_hi(sp.hi() + rustc_span::BytePos(whitespace as u32));
```
`whitespace` here is counted in `char`s, not bytes.

petrochenkov replied:
Are there any other suggestions that remove full lines?
There probably are, in that case would be better to share this logic with them.

clubby789 replied:
Given that the iterator takes only `ascii_whitespace`, I think it makes sense to treat the char count as byte count? I'll look for similar suggestions

estebank reviewed (commented):
Although this is a way to address this specific issue, it might be a better idea to modify `rustc_errors::json::DiagnosticSpan::from_span_full` to have this logic of checking "does this span end on the last char of a line, followed by `\n`? if extend the span by one byte". Doing it there *doesn't* change how the diff is shown in diagnostics, but aids rustfix not have to care about this *and* doesn't require us to address these kind of spans in *every* suggestion we produce.

clubby789 commented:
Doing that globally seems to break a lot of `.fixed` tests:
```diff,rs
diff --git a/tests/ui/binop/false-binop-caused-by-missing-semi.fixed b/tests/ui/binop/false-binop-caused-by-missing-semi.fixed
index b47372c9064..441fad80c8f 100644
--- a/tests/ui/binop/false-binop-caused-by-missing-semi.fixed
+++ b/tests/ui/binop/false-binop-caused-by-missing-semi.fixed
@@ -3,8 +3,7 @@ fn foo() {}
 fn main() {
     let mut y = 42;
     let x = &mut y;
-    foo();
-    *x = 0;  //~ ERROR invalid left-hand side of assignment
+    foo();    *x = 0;  //~ ERROR invalid left-hand side of assignment
     let _ = x;
     println!("{y}");
 }
diff --git a/tests/ui/block-result/consider-removing-last-semi.fixed b/tests/ui/block-result/consider-removing-last-semi.fixed
index 36a769fe529..931a5440b79 100644
--- a/tests/ui/block-result/consider-removing-last-semi.fixed
+++ b/tests/ui/block-result/consider-removing-last-semi.fixed
@@ -2,19 +2,16 @@
 
 pub fn f() -> String {  //~ ERROR mismatched types
     0u8;
-    "bla".to_string()
-}
+    "bla".to_string()}
 
 pub fn g() -> String {  //~ ERROR mismatched types
     "this won't work".to_string();
-    "removeme".to_string()
-}
+    "removeme".to_string()}
 
 pub fn macro_tests() -> u32 {  //~ ERROR mismatched types
     macro_rules! mac {
         () => (1);
     }
-    mac!()
-}
+    mac!()}```

estebank commented:
@clubby789 what happens if you only do the span extension if the replacement is ""?

clubby789 committed:
If suggestion would leave an empty line, delete it

clubby789 commented:
Catches a lot more cases 👀 Good idea!

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7670109603/job/20905520729) [(plain)](https://github.com/rust-lang/rust/commit/0b7081067ce0c7ab8cd9f887ca40f1f257b0e71a/checks/20905520729/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=rust-cloud-vms[bot]
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_2871f28f-4fef-4521-a719-02b5bbf2fdef
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=ac989dafb156589655421d755f9c2b0d4615eea1
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_2871f28f-4fef-4521-a719-02b5bbf2fdef
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_2871f28f-4fef-4521-a719-02b5bbf2fdef
GITHUB_TRIGGERING_ACTOR=rust-cloud-vms[bot]
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/120305/merge
GITHUB_WORKFLOW_SHA=ac989dafb156589655421d755f9c2b0d4615eea1
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
Built container sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Looks like docker image is the same as before, not uploading
https://ci-caches.rust-lang.org/docker/7ebc15c01a233894034d277c8cce4e949f4e7791f66b4727c8fb6e058a0b8171d6152e1441d677cef0653843ceeee469c097b8699b2bb74249e674f6aa1a8813
sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Fri Jan 26 15:35:24 UTC 2024
  network time: Fri, 26 Jan 2024 15:35:24 GMT
  network time: Fri, 26 Jan 2024 15:35:24 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'build.optimized-compiler-builtins', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
---- [ui] tests/ui/associated-types/impl-wf-cycle-6.rs stdout ----
diff of fixed:

19 
20 impl<T: Grault> Grault for (T,)
21 //~^ ERROR overflow evaluating the requirement `<(T,) as Grault>::A == _`
23 {
24     type A = ();
25     type B = bool;



The actual fixed differed from the expected fixed.
Actual fixed saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/associated-types/impl-wf-cycle-6/impl-wf-cycle-6.fixed
To only update this specific test, also pass `--test-args associated-types/impl-wf-cycle-6.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/associated-types/impl-wf-cycle-6.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/associated-types/impl-wf-cycle-6" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/associated-types/impl-wf-cycle-6/auxiliary"
--- stderr -------------------------------
--- stderr -------------------------------
error[E0275]: overflow evaluating the requirement `<(T,) as Grault>::A == _`
   |
   |
LL | / impl<T: Grault> Grault for (T,)
LL | | //~^ ERROR overflow evaluating the requirement `<(T,) as Grault>::A == _`
LL | | where
LL | |     Self::A: Baz,
LL |   {
LL |       type A = ();
LL |       type A = ();
   |       ------ associated type `<(T,) as Grault>::A` is specified here
   |
note: required for `(T,)` to implement `Grault`
   |
   |
LL | impl<T: Grault> Grault for (T,)
...
...
LL |     Self::A: Baz,
   |              --- unsatisfied trait bound introduced here
help: associated type for the current `impl` cannot be restricted in `where` clauses, remove this bound
LL - where
LL - where
LL -     Self::A: Baz,

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0275`.
For more information about this error, try `rustc --explain E0275`.
------------------------------------------


---- [ui] tests/ui/lint/unused/import_remove_line.rs stdout ----
diff of fixed:

6 
7 //~^ WARN unused imports
8 //~^ WARN unused import
+ 
9 //~^ WARN unused import
10 //~| WARN unused import


The actual fixed differed from the expected fixed.
Actual fixed saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/unused/import_remove_line/import_remove_line.fixed
---
status: exit status: 0
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/unused/import_remove_line.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/unused/import_remove_line" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/unused/import_remove_line/auxiliary"
stdout: none
--- stderr -------------------------------
warning: unused imports: `Duration`, `Instant`
   |
   |
LL | use std::time::{Duration, Instant};
   |
note: the lint level is defined here
  --> /checkout/tests/ui/lint/unused/import_remove_line.rs:5:9
   |
---

warning: unused import: `std::time::SystemTimeError`
##[warning]  --> /checkout/tests/ui/lint/unused/import_remove_line.rs:11:5
   |
LL | use std::time::SystemTimeError;use std::time::TryFromFloatSecsError;

warning: unused import: `std::time::TryFromFloatSecsError`
##[warning]  --> /checkout/tests/ui/lint/unused/import_remove_line.rs:11:36
   |
   |
LL | use std::time::SystemTimeError;use std::time::TryFromFloatSecsError;

warning: 4 warnings emitted
------------------------------------------

```

</details>

clubby789 commented:
Actually this could do with some improvement
@rustbot author

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7670943671/job/20908234639) [(plain)](https://github.com/rust-lang/rust/commit/ea182594a7cc61cad249b5f54306203198932175/checks/20908234639/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=rust-cloud-vms[bot]
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_536e2d68-479d-4658-aaa2-8a04f9e7bb82
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=5843394a7e3ea55c0e7b673a639485f5646d221c
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_536e2d68-479d-4658-aaa2-8a04f9e7bb82
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_536e2d68-479d-4658-aaa2-8a04f9e7bb82
GITHUB_TRIGGERING_ACTOR=rust-cloud-vms[bot]
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/120305/merge
GITHUB_WORKFLOW_SHA=5843394a7e3ea55c0e7b673a639485f5646d221c
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---

Location:
    /cargo/registry/src/index.crates.io-6f17d22bba15001f/ui_test-0.21.2/src/lib.rs:388:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Execute `cargo uibless` to update `tests/ui/derivable_impls.fixed` to the actual output
--- tests/ui/derivable_impls.fixed
+++ <fixed output>
 
... 17 lines skipped ...
 
 
---
 
 
-
 #[derive(Default)]
 struct AlreadyDerived(i32, bool);
... 18 lines skipped ...
 struct Y(u32);
-
 struct RustIssue26925<T> {
     a: Option<T>,
... 31 lines skipped ...
---

FAILED TEST: tests/ui/empty_drop.rs
command: "checking tests/ui/empty_drop.rs"

error: actual output differed from expected
Execute `cargo uibless` to update `tests/ui/empty_drop.fixed` to the actual output
--- tests/ui/empty_drop.fixed
+++ <fixed output>
 #![warn(clippy::empty_drop)]
... 3 lines skipped ...
 
 
-
---

FAILED TEST: tests/ui/must_use_unit.rs
command: "checking tests/ui/must_use_unit.rs"

error: actual output differed from expected
Execute `cargo uibless` to update `tests/ui/must_use_unit.fixed` to the actual output
--- tests/ui/must_use_unit.fixed
+++ <fixed output>
 //@aux-build:proc_macros.rs
... 4 lines skipped ...
 use proc_macros::external;
 
-
---

FAILED TEST: tests/ui/single_component_path_imports.rs
command: "checking tests/ui/single_component_path_imports.rs"

error: actual output differed from expected
Execute `cargo uibless` to update `tests/ui/single_component_path_imports.fixed` to the actual output
--- tests/ui/single_component_path_imports.fixed
+++ <fixed output>
 #![warn(clippy::single_component_path_imports)]
... 2 lines skipped ...
 
 
-
-
 use serde as edres;
... 31 lines skipped ...
... 31 lines skipped ...
     fn hi_mod() {}


full stderr:

---

error: test failed, to rerun pass `--test compile-test`


command did not execute successfully: cd "/checkout" && env -u MAKEFLAGS -u MFLAGS AR_x86_64_unknown_linux_gnu="ar" CARGO_INCREMENTAL="0" CARGO_PROFILE_RELEASE_DEBUG="0" CARGO_PROFILE_RELEASE_DEBUG_ASSERTIONS="true" CARGO_PROFILE_RELEASE_OVERFLOW_CHECKS="true" CARGO_PROFILE_RELEASE_STRIP="false" CARGO_TARGET_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools" CC_x86_64_unknown_linux_gnu="sccache cc" CFG_COMPILER_BUILD_TRIPLE="x86_64-unknown-linux-gnu" CFG_COMPILER_HOST_TRIPLE="x86_64-unknown-linux-gnu" CFG_RELEASE="1.77.0-nightly" CFG_RELEASE_CHANNEL="nightly" CFG_RELEASE_NUM="1.77.0" CFG_VERSION="1.77.0-nightly (5843394a7 2024-01-26)" CFG_VER_DATE="2024-01-26" CFG_VER_HASH="5843394a7e3ea55c0e7b673a639485f5646d221c" CFLAGS_x86_64_unknown_linux_gnu="-ffunction-sections -fdata-sections -fPIC -m64" CXXFLAGS_x86_64_unknown_linux_gnu="-ffunction-sections -fdata-sections -fPIC -m64" CXX_x86_64_unknown_linux_gnu="sccache c++" DOC_RUST_LANG_ORG_CHANNEL="https://doc.rust-lang.org/nightly" HOST_LIBS="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release" LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" LIBC_CHECK_CFG="1" LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm/lib" LIBZ_SYS_STATIC="1" LZMA_API_STATIC="1" RANLIB_x86_64_unknown_linux_gnu="ar s" REAL_LIBRARY_PATH_VAR="LD_LIBRARY_PATH" RUSTBUILD_NATIVE_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/native" RUSTC="/checkout/obj/build/bootstrap/debug/rustc" RUSTC_BOOTSTRAP="1" RUSTC_BREAK_ON_ICE="1" RUSTC_ERROR_METADATA_DST="/checkout/obj/build/tmp/extended-error-metadata" RUSTC_HOST_FLAGS="-Zunstable-options --check-cfg=cfg(bootstrap)" RUSTC_INSTALL_BINDIR="bin" RUSTC_LIBDIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" RUSTC_LIB_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" RUSTC_LINT_FLAGS="-Wrust_2018_idioms -Wunused_lifetimes -Dwarnings" RUSTC_REAL="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUSTC_SNAPSHOT="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUSTC_SNAPSHOT_LIBDIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" RUSTC_STAGE="2" RUSTC_SYSROOT="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" RUSTC_TEST_SUITE="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUSTC_TLS_MODEL_INITIAL_EXEC="1" RUSTC_VERBOSE="0" RUSTC_WRAPPER="/checkout/obj/build/bootstrap/debug/rustc" RUSTDOC="/checkout/obj/build/bootstrap/debug/rustdoc" RUSTDOCFLAGS="--cfg=windows_raw_dylib -Csymbol-mangling-version=v0 -Zunstable-options --check-cfg=cfg(bootstrap) --check-cfg=cfg(parallel_compiler) --check-cfg=cfg(rust_analyzer) --check-cfg=cfg(no_btreemap_remove_entry) --check-cfg=cfg(crossbeam_loom) --check-cfg=cfg(span_locations) --check-cfg=cfg(rustix_use_libc) --check-cfg=cfg(emulate_second_only_system) --check-cfg=cfg(windows_raw_dylib) -Dwarnings -Wrustdoc::invalid_codeblock_attributes --crate-version 1.77.0-nightly\t(5843394a7\t2024-01-26)" RUSTDOC_LIBDIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" RUSTDOC_REAL="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" RUSTFLAGS="--cfg=windows_raw_dylib -Csymbol-mangling-version=v0 -Zunstable-options --check-cfg=cfg(bootstrap) --check-cfg=cfg(parallel_compiler) --check-cfg=cfg(rust_analyzer) --check-cfg=cfg(no_btreemap_remove_entry) --check-cfg=cfg(crossbeam_loom) --check-cfg=cfg(span_locations) --check-cfg=cfg(rustix_use_libc) --check-cfg=cfg(emulate_second_only_system) --check-cfg=cfg(windows_raw_dylib) -Zmacro-backtrace -Clink-args=-Wl,-z,origin -Clink-args=-Wl,-rpath,$ORIGIN/../lib -Csplit-debuginfo=off -Zunstable-options" RUST_TEST_THREADS="16" SYSROOT="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" TERM="xterm" __CARGO_DEFAULT_LIB_METADATA="nightlytool-rustc" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--target" "x86_64-unknown-linux-gnu" "--release" "-Zcheck-cfg" "-Zbinary-dep-depinfo" "-j" "16" "--locked" "--color" "always" "--manifest-path" "/checkout/src/tools/clippy/Cargo.toml" "--" "--quiet"

stdout ----

stderr ----
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7680625630/job/20932893243) [(plain)](https://github.com/rust-lang/rust/commit/a1eda8f0a4334d698bf2882ac552cb2452b9c565/checks/20932893243/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=rust-cloud-vms[bot]
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_9aeda025-b79d-44ff-9468-4728ffbca1c8
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=7cbe418a289539ccf6fb2a114db4e3b03c911115
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_9aeda025-b79d-44ff-9468-4728ffbca1c8
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_9aeda025-b79d-44ff-9468-4728ffbca1c8
GITHUB_TRIGGERING_ACTOR=rust-cloud-vms[bot]
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/120305/merge
GITHUB_WORKFLOW_SHA=7cbe418a289539ccf6fb2a114db4e3b03c911115
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---

FAILED TEST: tests/ui/derivable_impls.rs
command: "checking tests/ui/derivable_impls.rs"

error: actual output differed from expected
Execute `cargo uibless` to update `tests/ui/derivable_impls.fixed` to the actual output
--- tests/ui/derivable_impls.fixed
+++ <fixed output>
called `Result::unwrap()` on an `Err` value: tests failed

Location:
    /cargo/registry/src/index.crates.io-6f17d22bba15001f/ui_test-0.21.2/src/lib.rs:388:13
---
 
 
-
 #[derive(Default)]
 struct AlreadyDerived(i32, bool);
... 18 lines skipped ...
... 18 lines skipped ...
 struct Y(u32);
-
 struct RustIssue26925<T> {
     a: Option<T>,
... 31 lines skipped ...
---

FAILED TEST: tests/ui/empty_drop.rs
command: "checking tests/ui/empty_drop.rs"

error: actual output differed from expected
Execute `cargo uibless` to update `tests/ui/empty_drop.fixed` to the actual output
--- tests/ui/empty_drop.fixed
+++ <fixed output>
 #![warn(clippy::empty_drop)]
... 3 lines skipped ...
 
 
-
---

FAILED TEST: tests/ui/must_use_unit.rs
command: "checking tests/ui/must_use_unit.rs"

error: actual output differed from expected
Execute `cargo uibless` to update `tests/ui/must_use_unit.fixed` to the actual output
--- tests/ui/must_use_unit.fixed
+++ <fixed output>
 //@aux-build:proc_macros.rs
... 4 lines skipped ...
 use proc_macros::external;
 
-
---

FAILED TEST: tests/ui/single_component_path_imports.rs
command: "checking tests/ui/single_component_path_imports.rs"

error: actual output differed from expected
Execute `cargo uibless` to update `tests/ui/single_component_path_imports.fixed` to the actual output
--- tests/ui/single_component_path_imports.fixed
+++ <fixed output>
 #![warn(clippy::single_component_path_imports)]
... 2 lines skipped ...
 
 
-
-
 use serde as edres;
... 31 lines skipped ...
... 31 lines skipped ...
     fn hi_mod() {}


full stderr:

---

Build completed unsuccessfully in 0:01:50


command did not execute successfully: cd "/checkout" && env -u MAKEFLAGS -u MFLAGS AR_x86_64_unknown_linux_gnu="ar" CARGO_INCREMENTAL="0" CARGO_PROFILE_RELEASE_DEBUG="0" CARGO_PROFILE_RELEASE_DEBUG_ASSERTIONS="true" CARGO_PROFILE_RELEASE_OVERFLOW_CHECKS="true" CARGO_PROFILE_RELEASE_STRIP="false" CARGO_TARGET_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools" CC_x86_64_unknown_linux_gnu="sccache cc" CFG_COMPILER_BUILD_TRIPLE="x86_64-unknown-linux-gnu" CFG_COMPILER_HOST_TRIPLE="x86_64-unknown-linux-gnu" CFG_RELEASE="1.77.0-nightly" CFG_RELEASE_CHANNEL="nightly" CFG_RELEASE_NUM="1.77.0" CFG_VERSION="1.77.0-nightly (7cbe418a2 2024-01-27)" CFG_VER_DATE="2024-01-27" CFG_VER_HASH="7cbe418a289539ccf6fb2a114db4e3b03c911115" CFLAGS_x86_64_unknown_linux_gnu="-ffunction-sections -fdata-sections -fPIC -m64" CXXFLAGS_x86_64_unknown_linux_gnu="-ffunction-sections -fdata-sections -fPIC -m64" CXX_x86_64_unknown_linux_gnu="sccache c++" DOC_RUST_LANG_ORG_CHANNEL="https://doc.rust-lang.org/nightly" HOST_LIBS="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release" LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" LIBC_CHECK_CFG="1" LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm/lib" LIBZ_SYS_STATIC="1" LZMA_API_STATIC="1" RANLIB_x86_64_unknown_linux_gnu="ar s" REAL_LIBRARY_PATH_VAR="LD_LIBRARY_PATH" RUSTBUILD_NATIVE_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/native" RUSTC="/checkout/obj/build/bootstrap/debug/rustc" RUSTC_BOOTSTRAP="1" RUSTC_BREAK_ON_ICE="1" RUSTC_ERROR_METADATA_DST="/checkout/obj/build/tmp/extended-error-metadata" RUSTC_HOST_FLAGS="-Zunstable-options --check-cfg=cfg(bootstrap)" RUSTC_INSTALL_BINDIR="bin" RUSTC_LIBDIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" RUSTC_LIB_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" RUSTC_LINT_FLAGS="-Wrust_2018_idioms -Wunused_lifetimes -Dwarnings" RUSTC_REAL="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUSTC_SNAPSHOT="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUSTC_SNAPSHOT_LIBDIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" RUSTC_STAGE="2" RUSTC_SYSROOT="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" RUSTC_TEST_SUITE="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUSTC_TLS_MODEL_INITIAL_EXEC="1" RUSTC_VERBOSE="0" RUSTC_WRAPPER="/checkout/obj/build/bootstrap/debug/rustc" RUSTDOC="/checkout/obj/build/bootstrap/debug/rustdoc" RUSTDOCFLAGS="--cfg=windows_raw_dylib -Csymbol-mangling-version=v0 -Zunstable-options --check-cfg=cfg(bootstrap) --check-cfg=cfg(parallel_compiler) --check-cfg=cfg(rust_analyzer) --check-cfg=cfg(no_btreemap_remove_entry) --check-cfg=cfg(crossbeam_loom) --check-cfg=cfg(span_locations) --check-cfg=cfg(rustix_use_libc) --check-cfg=cfg(emulate_second_only_system) --check-cfg=cfg(windows_raw_dylib) -Dwarnings -Wrustdoc::invalid_codeblock_attributes --crate-version 1.77.0-nightly\t(7cbe418a2\t2024-01-27)" RUSTDOC_LIBDIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" RUSTDOC_REAL="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" RUSTFLAGS="--cfg=windows_raw_dylib -Csymbol-mangling-version=v0 -Zunstable-options --check-cfg=cfg(bootstrap) --check-cfg=cfg(parallel_compiler) --check-cfg=cfg(rust_analyzer) --check-cfg=cfg(no_btreemap_remove_entry) --check-cfg=cfg(crossbeam_loom) --check-cfg=cfg(span_locations) --check-cfg=cfg(rustix_use_libc) --check-cfg=cfg(emulate_second_only_system) --check-cfg=cfg(windows_raw_dylib) -Zmacro-backtrace -Clink-args=-Wl,-z,origin -Clink-args=-Wl,-rpath,$ORIGIN/../lib -Csplit-debuginfo=off -Zunstable-options" RUST_TEST_THREADS="16" SYSROOT="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" TERM="xterm" __CARGO_DEFAULT_LIB_METADATA="nightlytool-rustc" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--target" "x86_64-unknown-linux-gnu" "--release" "-Zcheck-cfg" "-Zbinary-dep-depinfo" "-j" "16" "--locked" "--color" "always" "--manifest-path" "/checkout/src/tools/clippy/Cargo.toml" "--" "--quiet"

stdout ----

stderr ----
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #121810) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

estebank reviewed (approved):
please rebase and I'll approve

clubby789 commented:
@rustbot ready

estebank commented:
@bors r+

bors commented:
:pushpin: Commit 367126d49af3237eee24889191dfbb02dfc9320c has been approved by `estebank`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=estebank 367126d49af3237eee24889191dfbb02dfc9320c -->
<!-- homu: {"type":"Approved","sha":"367126d49af3237eee24889191dfbb02dfc9320c","approver":"estebank","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
